### ISSUE TITLE
Your setting of the conv dilation rate

### ISSUE THREAD

**Issue by Lebo-Wang** [philipperemy/keras-tcn#1]
Your conv_1d setting is:
`atrous_rate=2 ** i, i is from [1, 2, 4, 8]
`
That means you want to make the dilatations of conv1d as 
`atrous_rate=[2, 4, 16, 256]`

My sequence is 100 time-steps. As my guessing, the dilation_rate over 100 cannot be fully effective. But I tried:
`atrous_rate=2 ** i, i is from [1, 2, 3, 4, 5, 6]
`
or
`atrous_rate=2 ** i, i is from [1, 2, 4, 6]
`
Only your configuration with even dilate_rate=256 (>100) gives me the best result.

Can you explain more about this configuration? Thanks!

**Comment by philipperemy** [comment]
@Lebo-Wang I'll have to dig a bit more into this to explain this behavior.

But I think adding more than less is better. You always want to make sure that the receptive field covers all the sequence.

What you can also do as a test (I also plan to do it) is to train on a sequence like this:

x = `[1.0, 0.0, 0.0, 0.0, 0.0, ...., 0.0, 0.0, 0.0, 0.0]` y = `[1.0]`
x = `[0.0, 0.0, 0.0, 0.0, 0.0, ...., 0.0, 0.0, 0.0, 0.0]` y = `[0.0]`

Where x has length 100. You will clearly see if your receptive field is long enough because the network has to understand the relation `y = x[0]`. So it must be able to see at the beginning of the long sequence.

**Comment by Lebo-Wang** [comment]
@philipperemy Thanks for the reply! I think you are always right to make the receptive field larger. But according to my understanding and reference, e.g:
`dilations = [2**i for i in range(N)] * M`
[https://github.com/ibab/tensorflow-wavenet/blob/e11ad19dccd6a33a182c9d1dea07aa53b9acca55/wavenet/model.py#L35](url)
Also, if the receptive field is larger than the sequence length, I guess this overflow field will never be fully utilized. Of course, the shortcut path will still receive information from previous layers within the sequence length.

**Comment by Lebo-Wang** [comment]
@philipperemy I keep testing all kinds of dilation_rate combinations on my own dataset, whose length is 100. The only difference for the architecture is that I return the classification result in every step and get the average result for the final classification (instead of only depending on the last one result). Also, I use Batchnorm layers to make the model easier to train. Here are some results:

1. The performance of the `atrous_rate=[2, 4, 16, 256]` is quite close to `atrous_rate=[2, 4, 16, 2]`. The reason is that the 256 dilated Conv is beyond my sequence length (100), which is no more than one stacked layer with 1*1 Conv.

2. Strictly speaking, the receptive field of the `atrous_rate=[2, 4, 16, 256]` is actually 255. I show the `atrous_rate=[2, 4, 16]`, which is 15:
![blank diagram - page 2 2](https://user-images.githubusercontent.com/26159842/38445349-0b5d2bfa-39a8-11e8-895d-875267e4f071.png)

3. The performance of different combinations (Maybe not fully trained):
`atrous_rate=[0, 2, 4, 16] > atrous_rate=[2, 4, 16, 2] > atrous_rate=[2, 4, 16, 256]`

**Comment by philipperemy** [comment]
@Lebo-Wang very interesting.

Do you think it would be more reasonable to have:

```
dilations = [2**i for i in range(N)] * M
```

**Comment by Lebo-Wang** [comment]
@philipperemy Yes, I think so. Most of the TCN is like this, where 2 is the kernel size.

**Comment by philipperemy** [comment]
Ok I'm going to implement that. Thanks for the feedback. Highly appreciated :)

**Comment by philipperemy** [comment]
Done in https://github.com/philipperemy/keras-tcn/commit/b6bec26c74ec692d0dccc3e487465002fbb8ad7c

Thanks a lot :)

And sorry for the delay!
